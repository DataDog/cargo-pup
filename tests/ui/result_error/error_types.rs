//@compile-flags: --crate-name test
//@compile-flags: --crate-type lib

mod test {
    use std::fmt;

    // Custom error type that doesn't implement std::error::Error
    struct CustomError {
        message: String,
    }

    // For Display formatting, but not implementing std::error::Error
    impl fmt::Display for CustomError {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            write!(f, "{}", self.message)
        }
    }

    // Custom error type that properly implements std::error::Error
    #[derive(Debug)]
    struct GoodCustomError {
        message: String,
    }

    impl std::error::Error for GoodCustomError {}

    impl fmt::Display for GoodCustomError {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            write!(f, "{}", self.message)
        }
    }

    // This function should trigger a lint error - i32 doesn't implement Error
    fn bad_primitive_error() -> Result<(), i32> { //~ ERROR: Type 'i32' is used as an error type in Result but does not implement Error trait
        Ok(())
    }

    // This function should trigger a lint error - CustomError doesn't implement Error
    fn bad_custom_error() -> Result<(), CustomError> { //~ ERROR: Type 'test::CustomError' is used as an error type in Result but does not implement Error trait
        Ok(())
    }

    // This function should be allowed - String implements Error
    fn good_string_error() -> Result<(), String> {
        Ok(())
    }

    // This function should be allowed - std::io::Error implements Error
    fn good_io_error() -> Result<(), std::io::Error> {
        Ok(())
    }

    // This function should be allowed - GoodCustomError implements Error
    fn good_custom_error() -> Result<(), GoodCustomError> {
        Ok(())
    }
}