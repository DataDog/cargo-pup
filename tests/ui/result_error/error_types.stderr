error: Type 'i32' is used as an error type in Result but does not implement Error trait
  --> tests/ui/result_error/error_types.rs:34:5
   |
LL | /     fn bad_primitive_error() -> Result<(), i32> {
LL | |         Ok(())
LL | |     }
   | |_____^
   |
   = help: Implement the Error trait for this type or use a type that implements Error.
   = note: Applied by cargo-pup rule 'result_error_test'.
   = note: `#[deny(result_error)]` on by default

error: Type 'test::CustomError' is used as an error type in Result but does not implement Error trait
  --> tests/ui/result_error/error_types.rs:39:5
   |
LL | /     fn bad_custom_error() -> Result<(), CustomError> {
LL | |         Ok(())
LL | |     }
   | |_____^
   |
   = help: Implement the Error trait for this type or use a type that implements Error.
   = note: Applied by cargo-pup rule 'result_error_test'.

error: aborting due to 2 previous errors

